### DISCUSSION TITLE
[10주차] 네트워크 보안 8.1~4

### DISCUSSION THREAD

**Discussion by Hanjaemo** [Dijkstra-Employ/Network-Study#17] [Q&A]
10주차 네트워크 보안 8.1~4에 해당하는 질의응답입니다.

**Comment by Starlight258** [comment]
# Q1. 
오늘날 사용되는 `블록 암호화 방법`에는 무엇이 있나요?

# Q2.
`RSA`와 `세션키`를 함께 사용할 때, `세션키`의 사용 목적은 무엇인가요?

**Reply by H-sooyeon** [reply]
> Q1 오늘날 사용되는 블록 암호화 방법에는 무엇이 있나요?

블록 암호는 고전된 n비트 블록을 n비트 블록으로 변환시키는 함수이다.
즉, 임의 길이의 평문을 암호화시키기 위해서는 평문을 특정한 길이로 분할시켜야 한다.
DES나 AES, SEED가 대표적이다.
- DES: 블록 크기는 64비트, 키 길이는 56비트이며 다량의 문서를 암호화할 수 있다.
- AES: 블록 크기는 128비트, 평문을 128비트 암호문으로 출력하며 높은 안정성과 속도를 가지고 있다. 랜섬웨어에 많이 사용된다.
- SEED: 인터넷, 전자상거래, 무선통신 등에 사용하기 위해 만들어진 알고리즘이며 블록 크기는 128비트이다.

<br />

> Q2 RSA와 세션키를 함께 사용할 때, 세션키의 사용 목적은 무엇인가요?

대칭 암호화 방식을 통해 데이터 전송을 효율적이고 빠르게 암호화하기 위해서이다.
RSA는 공개키 기반의 비대칭 암호화 방식으로, 보안성이 뛰어나지만 속도가 느려 대량의 데이터를 암호화하기에는 비효율적이다.
이러한 문제를 해결하기 위해 세션키를 사용한다.
또한, 세션마다 새로운 키가 사용되기 때문에 한 번의 키가 노출되더라도 다른 세션에는 영향을 미치지 않는다.

**Reply by Starlight258** [reply]
> # Q1.
> 오늘날 사용되는 `블록 암호화 방법`에는 무엇이 있나요?
> 
- `DES`, `3DES`, `AES` 등이 있다.
- 각 알고리즘들은 비트열을 키로 사용한다.
  - DES는 56비트의 키를 64비트 블록과 함께 사용한다.
  - 알고리즘의 키는 알고리즘 내부에서 특정 미니 테이블의 사상과 순열을 결정한다.

> # Q2.
> `RSA`와 `세션키`를 함께 사용할 때, `세션키`의 사용 목적은 무엇인가요?
- RSA는 시간이 많이 필요하므로, 성능을 위해 대칭키 암호화와 함께 사용하기도 한다.
- 동작
	- 앨리스는 데이터 암호화에 사용할 키(세션키)를 고른다
	- 앨리스는 밥의 공개키로 세션키를 암호화한다.
	- 밥은 개인키로 복호화하여 세션키를 얻는다.
- RSA로 세션키를 교환한 후 세션키로 통신한다.

**Comment by H-sooyeon** [comment]
## Q1.
대칭키 암호화와 공개키 암호화에 대해 설명해주세요.

## Q2.
메시지 무결성이 필요한 이유와, 전자서명이 어떻게 무결성을 보장해주는 것인지 설명해주세요.

**Reply by Starlight258** [reply]
> ## Q1.
> 대칭키 암호화와 공개키 암호화에 대해 설명해주세요.
> 
- **대칭키 암호화** : 두 종단 시스템의 사용자가 가진 키가 동일하다.
- **공개키 암호화**: 한 쌍의 키가 사용되며, 한 키는 공개키이고 나머지 하나는 비밀키이다.

> ## Q2.
> 메시지 무결성이 필요한 이유와, 전자서명이 어떻게 무결성을 보장해주는 것인지 설명해주세요.
- 메세지 무결성은 해당 메세지를 보낸 사람의 신원이 보증되는지, 전송되는중에 메세지가 변경되지는 않았는지 검증하는 역할을 한다.

전자서명
- 송신자가 개인키로 암호화(전자서명) 후 수신자는 송신자의 공개키로 복호화를 수행한다.
- 복호화가 성공적으로 수행될 경우, 수신자에게 송신자의 신분 뿐 아니라 메세지가 변형되지 않음을 확인할 수 있다.

**Comment by choiyoung69** [comment]
## Q1
MAC이 메세지 무결성을 보장해주는 방법을 설명해주세요.

## Q2
인증 프로토콜 ap 1.0, 2.0, 3.0의 문제점과 이 문제점을 해결한 ap 4.0에 대해 설명해주세요

**Reply by H-sooyeon** [reply]
> Q1 MAC이 메세지 무결성을 보장해주는 방법을 설명해주세요.

![image](https://github.com/user-attachments/assets/4b033b02-f3f6-469b-ad6e-b24d6df68ac0)
[대칭 암호 사용]

![image](https://github.com/user-attachments/assets/d505a3b6-d51e-4cb7-a774-cc984615becf)
[공개키 암호 사용]

MAC(메시지 인증 코드)는 메시지 무결성을 보장하고, 메시지가 중간에 변경되지 않았음을 확인하는 데 사용되는 암호화 기술이다.
- 송신자는 메시지`(m)`와 함께 비밀 키`(k)`를 사용해 MAC`(MAC(K, M))`을 생성한다.
- 송신자는 원본 메시지`(m)`과 함께 MAC`(MAC(K, M))` 값을 수신자에게 전송한다.
- 수신자가 동일한 MAC 값을 생성할 수 있는 유일한 조건은 메시지가 변경되지 않았고, 동일한 비밀 키를 공유하고 있을 때이다.
- 수신자는 메시지를 받은 후, 자신이 보유한 비밀 키(k)와 받은 메시지(m)를 사용하여 MAC을 다시 생성한다.
- 수신자가 생성한 MAC 값과 송신자로부터 받은 MAC 값을 비교한다.
- 두 MAC 값이 일치하면, 메시지가 전송 과정에서 변경되지 않았다는 무결성을 보장하게 된다.
- 만약 MAC 값이 다르면, 메시지가 중간에 변조되었거나 무결성이 손상된 것으로 판단할 수 있다.

<br />

> Q2 인증 프로토콜 ap 1.0, 2.0, 3.0의 문제점과 이 문제점을 해결한 ap 4.0에 대해 설명해주세요

ap1.0의 문제점은 비밀번호와 같은 중요한 인증 정보를 암호화하지 않고 평문으로 전송한다는 점이다.
이는 중간자 공격(중간에서 정보 탈취) 및 재전송 공격에 취약했다.

ap2.0의 문제점은 평문 전송의 문제를 해결하기 위해 데이터를 암호화했으나, 사용된 암호화 알고리즘이 보안에 약한 방식이었다는 것이다. 이는 암호 분석 공격 및 세션 하이재킹 공격에 취약했다.

ap3.0은 키 교환 과정에서 중간자 공격의 가능성이 있었으며, 서버에 대한 인증이 부족했다.(상호 인증 부족)
그렇기에 공격자가 가짜 서버를 만들어 사용자의 정보를 탈취할 수 있는 피싱 공격에 취약했다.

ap4.0에서는 이러한 문제점들을 해결하기 위해 공개키 기반의 암호화를 사용하여 안전하게 키를 교환하고, CA를 통한 상호 인증으로 피싱 공격을 방지하였다. 또한, AES같은 강력한 암호화 알고리즘을 사용하여 데이터를 보호하여 다양한 보안 문제점들을 보완했다.

**Reply by Starlight258** [reply]
> ## Q1
> MAC이 메세지 무결성을 보장해주는 방법을 설명해주세요.
> 
- 0. 송신자와 수신자는 인증키(공유 비밀키, s)를 서로 공유한다.
- 1. 앨리스는 메세지 m을 생성하고 해시 `H(m+s)`을 계산한다. (`H(m+s)`를 `MAC`이라고 부른다).
- 2. 앨리스는 메세지 m에 MAC을 첨부하여 확장 메세지 `(m, H(m+s))`을 생성한 후 밥에게 보낸다.
- 3. `확장 메세지 (m, h)`를 받은 밥은 `MAC`을 계산한다. 만일 `H(m+s) = h`라면, 밥은 무결성이 훼손되지 않음을 알 수 있다.

> ## Q2
> 인증 프로토콜 ap 1.0, 2.0, 3.0의 문제점과 이 문제점을 해결한 ap 4.0에 대해 설명해주세요
- ap 1.0 문제점 : 
   - 앨리스가 **자신이 앨리스라고 언급한 메세지**(신원에 대한 정보)를 밥에게 보낸다.
   - 침입자 또한 같은 메세지를 보낼 수 있으므로 안전하지 않다.
- ap 2.0 문제점 : 
  - 밥은 인증 메세지의 IP 데이터그램의 **출발 주소가 앨리스의 IP 주소**인지 확인하여 인증한다.
  - `IP 스푸핑`을 통해 침입자가 앨리스의 주소를 IP 데이터그램의 출발주소로 하여 전송할 수 있으므로 안전하지 않다.
- ap 3.0 문제점 : 
  - **비밀번호**를 사용하여 서로를 인증한다.
  - 침입자는 앨리스의 통신을 도청하여 앨리스의 비밀번호를 알아낼 수 있어 안전하지 않다.
- ap 3.1 문제점 : 
  - **비밀번호를 암호화**하여 메세지와 함께 전송한다.
  - 트루디는 앨리스의 통신을 도청하여 암호화된 비밀번호를 저장하고, 후에 밥에게 앨리스인것처럼 메세지를 전송할 수 있다. (**재생 공격**)
- ap 4.0 해결 : 
  - **`넌스`(nonce) 를 사용하여 실시간성을 보장**한다.
	- `nonce` : 프로토콜이 평생에 단 한번 사용하는 숫자
	- 앨리스가 메세지를 보내면 밥은 `넌스`를 전송하고, 앨리스는 대칭 비밀키를 이용해 넌스를 암호화하여 전송한다.
	- 밥은 암호화한 넌스를 복호화하여 자신이 보낸 넌스와 같다면 앨리스를 인증한다. 
	- 앨리스가 실제 통신중이라는 것을 확인할 수 있다.

**Comment by Hanjaemo** [comment]
### Q1
HTTPS에서 대칭키와 공개키 암호화의 역할은 무엇인가요?

### Q2
대칭키 암호화가 적합한 경우와 부적합한 경우는 어떤 상황인가요?

**Reply by Starlight258** [reply]
> ### Q1
> HTTPS에서 대칭키와 공개키 암호화의 역할은 무엇인가요?
> 
- HTTPS는 SSL 프로토콜 위에서 동작하는 HTTP 프로토콜을 말한다.
- SSL 인증서는 CA의 정보와 서버측 공개키의 내용이 CA의 개인키로 서명되어있다.
  - 브라우저는 CA의 공개키로 복호화하여 신뢰성 있는 사이트임을 검증한다.
- SSL 핸드쉐이크 과정에서 서버의 신뢰성을 검증하고 서버와 클라이언트가 각각 제공한 랜덤 데이터를 통해 클라이언트는 pms를 생성한다.
  - 클라이언트는 psm를 서버의 공개키로 암호화하여 전달하고, 서버는 개인키로 복호화하여 pms를 공유한다. 이를 통해 master secret을 만들고 ms를 이용해 session key를 만든다.
  - 서버와 클라이언트가 데이터를 주고받는 과정에는 세션키를 이용한다.
    - 공개키 암호화는 컴퓨팅 파워를 크게 소모하므로 세션키를 공유하는 과정에서만 사용하고, 데이터 전달시에는 대칭키 방식을 사용한다.    

> ### Q2
> 대칭키 암호화가 적합한 경우와 부적합한 경우는 어떤 상황인가요?

대칭키 암호화 적합한 경우
- 빠른 속도의 암호화/복호화가 필요할 경우
- 구현이 쉽고 컴퓨팅 자원을 적게 소모해야할 경우

부적합한 경우
- 키를 안전하게 전달하려면 별도의 보안 채널이 필요하기 때문에 보안상 침해가 발생할 수 있다. (보안이 중요한 경우)
- 키 관리가 복잡하고 참여자 수만큼 대칭키가 존재해야한다. (키 관리가 쉽지 않음)

비대칭 암호화 적합한 경우
- 보안상 키 배포를 안전하게 수행해야할 경우
- 키 관리가 간단하고 효율적이다 

비대칭 암호화 부적합한 경우
- 암호화/복호화 과정이 빠르게 수행되어야할 경우
- 컴퓨팅 자원 소모량이 적어야 할 경우
- 알고리즘이 복잡하고 구현 및 유지보수가 어렵다.

HTTPS
- 공개키로 안전하게 세션 키를 교환
- 세션 키(대칭키)로 빠르게 데이터를 암호화/복호화
